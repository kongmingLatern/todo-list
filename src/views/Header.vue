<template>
  <div class="header">
    <div class="header-container">
      TodoList
      <input
      v-focus
      data-test="header-input"
      v-model="inputValue"
      @keyup.enter="addTodoItem"
      class="header-input"
      placeholder="todoItem"
      />
    </div>
  </div>
</template>

<script setup lang='ts'>
import { ref, defineEmits } from 'vue'

const inputValue = ref<string>('')

const emit = defineEmits<{
  (e: 'add', value: string): void
}>()

const addTodoItem = () => {
  if (inputValue.value) {
    emit('add', inputValue.value)
    inputValue.value = ''
  }
}

const vFocus = {
  mounted(el: any): void {
    // 获取input，并调用其focus()方法
    el.focus()
  }
}
</script>

<style lang="less" scoped>
.header {
  line-height: 60px;
  background: #333;
}

.header-container {
  width: 600px;
  margin: 0 auto;
  color: #fff;
  font-size: 24px;
}

.header-input {
  float: right;
  width: 360px;
  margin-top: 16px;
  line-height: 24px;
  outline: none;
  color: #333;
  text-indent: 10px;
}

input::placeholder {
  color: #999;
}
</style>
