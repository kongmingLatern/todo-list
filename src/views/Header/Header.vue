<template>
  <input data-test="input" />
</template>

<script setup lang='ts'>

</script>

<style>
</style>
