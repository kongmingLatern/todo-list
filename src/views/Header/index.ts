export default function add (a, b): number {
  return a + b
}
