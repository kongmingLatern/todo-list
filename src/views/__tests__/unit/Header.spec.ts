import { shallowMount } from '@vue/test-utils'
import Header from '@/views/Header.vue'

let wrapper: any = null
let input: any = null

beforeAll(() => {
  wrapper = shallowMount(Header)
  input = wrapper.find('[data-test="input"]')
})

describe('notice', () => {
  it('we should notice when developer changed something', () => {
    const wrapper = shallowMount(Header)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('input', () => {
  // 首先确保有这个 input 元素
  it('we should have a input Element', () => {
    expect(input.exists()).toBe(true)
  })

  // 获取到 input 的 value 值
  it('get input value', () => {
    // 通过 DOM 获取的值
    const value = wrapper.get('input').wrapperElement.value
    expect(value).toBe('')
  })

  // 当用户输入值后 input 的 value 值也应当同时改变
  it("input's value should be changed when user input the value", () => {
    // 假设用户输入了一串值
    input.setValue('this is test value')
    // 获取到用户输入的数据
    const value = wrapper.get('input').wrapperElement.value
    expect(value).toBe('this is test value')
  })

  // 当用户没有输入内容时输入回车，则不需要做任何操作
  it("we should't do anything when user just input keyEnter", () => {
    // 输入空内容
    input.setValue('')
    // 然后输入回车
    input.trigger('keyup.enter')
    expect(wrapper.emitted().add).toBeFalsy()
  })

  // 如果输入有内容，并且输入回车，则触发 emit 事件，同时清空 inputValue
  it('we should trigger emit and then clear inputValue when user input key is pressed', () => {
    const value = wrapper.get('input').wrapperElement.value
    input.setValue('anything')
    input.trigger('keyup.enter')
    expect(wrapper.emitted().add).toBeTruthy()
    expect(wrapper.emitted('add')[0]).toEqual(['anything'])
    expect(value).toBe('')
  })
})
