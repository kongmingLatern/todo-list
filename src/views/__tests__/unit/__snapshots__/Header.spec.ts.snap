// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notice we should notice when developer changed something 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <div
        class="header"
      >
        <div
          class="header-container"
        >
           TodoList 
          <input
            class="header-input"
            data-test="input"
            placeholder="todoItem"
          />
        </div>
      </div>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
        },
        "transition-group": Object {
          "name": "transition-group",
        },
      },
      "config": Object {
        "compilerOptions": Object {},
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "emitsCache": WeakMap {},
      "mixins": Array [
        Object {
          "beforeCreate": [Function],
        },
      ],
      "optionsCache": WeakMap {},
      "propsCache": WeakMap {},
      "provides": Object {},
      "reload": [Function],
    },
    "_instance": Object {
      "a": null,
      "accessCache": Object {},
      "appContext": Object {
        "app": [Circular],
        "components": Object {
          "transition": Object {
            "name": "transition",
          },
          "transition-group": Object {
            "name": "transition-group",
          },
        },
        "config": Object {
          "compilerOptions": Object {},
          "errorHandler": undefined,
          "globalProperties": Object {},
          "isNativeTag": [Function],
          "optionMergeStrategies": Object {},
          "performance": false,
          "warnHandler": undefined,
        },
        "directives": Object {},
        "emitsCache": WeakMap {},
        "mixins": Array [
          Object {
            "beforeCreate": [Function],
          },
        ],
        "optionsCache": WeakMap {},
        "propsCache": WeakMap {},
        "provides": Object {},
        "reload": [Function],
      },
      "asyncDep": null,
      "asyncResolved": false,
      "attrs": Object {},
      "bc": null,
      "bm": null,
      "bu": null,
      "bum": null,
      "c": null,
      "components": null,
      "ctx": Object {},
      "da": null,
      "data": Object {},
      "directives": null,
      "ec": null,
      "effect": ReactiveEffect {
        "active": true,
        "allowRecurse": true,
        "deps": Array [
          Set {
            [Circular],
          },
          Set {
            [Circular],
          },
          Set {
            [Circular],
          },
        ],
        "fn": [Function],
        "onTrack": undefined,
        "onTrigger": undefined,
        "parent": undefined,
        "scheduler": [Function],
      },
      "emit": [Function],
      "emitsOptions": null,
      "emitted": null,
      "exposeProxy": null,
      "exposed": null,
      "inheritAttrs": undefined,
      "isDeactivated": false,
      "isMounted": true,
      "isUnmounted": false,
      "m": null,
      "next": null,
      "parent": null,
      "props": Object {},
      "propsDefaults": Object {},
      "propsOptions": Array [
        Object {},
        Array [],
      ],
      "provides": Object {},
      "proxy": Object {},
      "refs": Object {
        "VTU_COMPONENT": Object {
          "hasOwnProperty": [Function],
        },
      },
      "render": [Function],
      "renderCache": Array [],
      "root": [Circular],
      "rtc": null,
      "rtg": null,
      "scope": EffectScope {
        "active": true,
        "cleanups": Array [],
        "effects": Array [
          ReactiveEffect {
            "active": true,
            "allowRecurse": true,
            "deps": Array [
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
            ],
            "fn": [Function],
            "onTrack": undefined,
            "onTrigger": undefined,
            "parent": undefined,
            "scheduler": [Function],
          },
        ],
      },
      "setupContext": null,
      "setupState": Object {},
      "slots": Object {},
      "sp": null,
      "subTree": Object {
        "__v_isVNode": true,
        "__v_skip": true,
        "anchor": null,
        "appContext": null,
        "children": null,
        "component": Object {
          "a": null,
          "accessCache": Object {},
          "appContext": Object {
            "app": [Circular],
            "components": Object {
              "transition": Object {
                "name": "transition",
              },
              "transition-group": Object {
                "name": "transition-group",
              },
            },
            "config": Object {
              "compilerOptions": Object {},
              "errorHandler": undefined,
              "globalProperties": Object {},
              "isNativeTag": [Function],
              "optionMergeStrategies": Object {},
              "performance": false,
              "warnHandler": undefined,
            },
            "directives": Object {},
            "emitsCache": WeakMap {},
            "mixins": Array [
              Object {
                "beforeCreate": [Function],
              },
            ],
            "optionsCache": WeakMap {},
            "propsCache": WeakMap {},
            "provides": Object {},
            "reload": [Function],
          },
          "asyncDep": null,
          "asyncResolved": false,
          "attrs": Object {},
          "bc": null,
          "bm": null,
          "bu": null,
          "bum": null,
          "c": null,
          "components": Object {},
          "ctx": Object {},
          "da": null,
          "data": Object {},
          "devtoolsRawSetupState": Object {
            "addTodoItem": [Function],
            "emit": [Function],
            "inputValue": RefImpl {
              "__v_isRef": true,
              "__v_isShallow": false,
              "_rawValue": "",
              "_value": "",
              "dep": Set {
                ReactiveEffect {
                  "active": true,
                  "allowRecurse": true,
                  "deps": Array [
                    [Circular],
                  ],
                  "fn": [Function],
                  "onTrack": undefined,
                  "onTrigger": undefined,
                  "parent": undefined,
                  "scheduler": [Function],
                },
              },
            },
          },
          "directives": null,
          "ec": null,
          "effect": ReactiveEffect {
            "active": true,
            "allowRecurse": true,
            "deps": Array [
              Set {
                [Circular],
              },
            ],
            "fn": [Function],
            "onTrack": undefined,
            "onTrigger": undefined,
            "parent": undefined,
            "scheduler": [Function],
          },
          "emit": [Function],
          "emitsOptions": Object {
            "add": null,
          },
          "emitted": null,
          "exposeProxy": Object {
            "hasOwnProperty": [Function],
          },
          "exposed": Object {
            "hasOwnProperty": [Function],
          },
          "inheritAttrs": undefined,
          "isDeactivated": false,
          "isMounted": true,
          "isUnmounted": false,
          "m": null,
          "next": null,
          "parent": [Circular],
          "props": Object {},
          "propsDefaults": Object {},
          "propsOptions": Array [
            Object {},
            Array [],
          ],
          "provides": Object {},
          "proxy": Object {},
          "refs": Object {},
          "render": [Function],
          "renderCache": Array [
            [Function],
          ],
          "root": [Circular],
          "rtc": null,
          "rtg": null,
          "scope": EffectScope {
            "active": true,
            "cleanups": Array [],
            "effects": Array [
              ReactiveEffect {
                "active": true,
                "allowRecurse": true,
                "deps": Array [
                  Set {
                    [Circular],
                  },
                ],
                "fn": [Function],
                "onTrack": undefined,
                "onTrigger": undefined,
                "parent": undefined,
                "scheduler": [Function],
              },
            ],
          },
          "setupContext": Object {
            "attrs": Object {},
            "emit": [Function],
            "expose": [Function],
            "slots": Object {},
          },
          "setupState": Object {
            "addTodoItem": [Function],
            "emit": [Function],
            "inputValue": "",
          },
          "slots": Object {},
          "sp": null,
          "subTree": Object {
            "__v_isVNode": true,
            "__v_skip": true,
            "anchor": null,
            "appContext": null,
            "children": Array [
              Object {
                "__v_isVNode": true,
                "__v_skip": true,
                "anchor": null,
                "appContext": null,
                "children": Array [
                  Object {
                    "__v_isVNode": true,
                    "__v_skip": true,
                    "anchor": null,
                    "appContext": null,
                    "children": " TodoList ",
                    "component": null,
                    "dirs": null,
                    "dynamicChildren": null,
                    "dynamicProps": null,
                    "el":  TodoList ,
                    "key": null,
                    "patchFlag": 0,
                    "props": null,
                    "ref": null,
                    "scopeId": null,
                    "shapeFlag": 8,
                    "slotScopeIds": null,
                    "ssContent": null,
                    "ssFallback": null,
                    "staticCount": 0,
                    "suspense": null,
                    "target": null,
                    "targetAnchor": null,
                    "transition": null,
                    "type": Symbol(Text),
                  },
                  Object {
                    "__v_isVNode": true,
                    "__v_skip": true,
                    "anchor": null,
                    "appContext": null,
                    "children": null,
                    "component": null,
                    "dirs": Array [
                      Object {
                        "arg": undefined,
                        "dir": Object {
                          "beforeUpdate": [Function],
                          "created": [Function],
                          "mounted": [Function],
                        },
                        "instance": Object {
                          "hasOwnProperty": [Function],
                        },
                        "modifiers": Object {},
                        "oldValue": undefined,
                        "value": "",
                      },
                    ],
                    "dynamicChildren": null,
                    "dynamicProps": Array [
                      "onKeyup",
                    ],
                    "el": <input
                      class="header-input"
                      data-test="input"
                      placeholder="todoItem"
                    />,
                    "key": null,
                    "patchFlag": 40,
                    "props": Object {
                      "class": "header-input",
                      "data-test": "input",
                      "onKeyup": [Function],
                      "onUpdate:modelValue": [Function],
                      "placeholder": "todoItem",
                    },
                    "ref": null,
                    "scopeId": null,
                    "shapeFlag": 1,
                    "slotScopeIds": null,
                    "ssContent": null,
                    "ssFallback": null,
                    "staticCount": 0,
                    "suspense": null,
                    "target": null,
                    "targetAnchor": null,
                    "transition": null,
                    "type": "input",
                  },
                ],
                "component": null,
                "dirs": null,
                "dynamicChildren": null,
                "dynamicProps": null,
                "el": <div
                  class="header-container"
                >
                   TodoList 
                  <input
                    class="header-input"
                    data-test="input"
                    placeholder="todoItem"
                  />
                </div>,
                "key": null,
                "patchFlag": 0,
                "props": Object {
                  "class": "header-container",
                },
                "ref": null,
                "scopeId": null,
                "shapeFlag": 17,
                "slotScopeIds": null,
                "ssContent": null,
                "ssFallback": null,
                "staticCount": 0,
                "suspense": null,
                "target": null,
                "targetAnchor": null,
                "transition": null,
                "type": "div",
              },
            ],
            "component": null,
            "dirs": null,
            "dynamicChildren": Array [
              Object {
                "__v_isVNode": true,
                "__v_skip": true,
                "anchor": null,
                "appContext": null,
                "children": null,
                "component": null,
                "dirs": Array [
                  Object {
                    "arg": undefined,
                    "dir": Object {
                      "beforeUpdate": [Function],
                      "created": [Function],
                      "mounted": [Function],
                    },
                    "instance": Object {
                      "hasOwnProperty": [Function],
                    },
                    "modifiers": Object {},
                    "oldValue": undefined,
                    "value": "",
                  },
                ],
                "dynamicChildren": null,
                "dynamicProps": Array [
                  "onKeyup",
                ],
                "el": <input
                  class="header-input"
                  data-test="input"
                  placeholder="todoItem"
                />,
                "key": null,
                "patchFlag": 40,
                "props": Object {
                  "class": "header-input",
                  "data-test": "input",
                  "onKeyup": [Function],
                  "onUpdate:modelValue": [Function],
                  "placeholder": "todoItem",
                },
                "ref": null,
                "scopeId": null,
                "shapeFlag": 1,
                "slotScopeIds": null,
                "ssContent": null,
                "ssFallback": null,
                "staticCount": 0,
                "suspense": null,
                "target": null,
                "targetAnchor": null,
                "transition": null,
                "type": "input",
              },
            ],
            "dynamicProps": null,
            "el": <div
              class="header"
            >
              <div
                class="header-container"
              >
                 TodoList 
                <input
                  class="header-input"
                  data-test="input"
                  placeholder="todoItem"
                />
              </div>
            </div>,
            "key": null,
            "patchFlag": 0,
            "props": Object {
              "class": "header",
            },
            "ref": null,
            "scopeId": null,
            "shapeFlag": 17,
            "slotScopeIds": null,
            "ssContent": null,
            "ssFallback": null,
            "staticCount": 0,
            "suspense": null,
            "target": null,
            "targetAnchor": null,
            "transition": null,
            "type": "div",
          },
          "suspense": null,
          "suspenseId": 0,
          "type": Object {
            "__name": "Header",
            "components": Object {},
            "emits": Array [
              "add",
            ],
            "render": [Function],
            "setup": [Function],
          },
          "u": null,
          "uid": 3,
          "um": null,
          "update": [Function],
          "vnode": [Circular],
          "withProxy": null,
        },
        "dirs": null,
        "dynamicChildren": null,
        "dynamicProps": null,
        "el": <div
          class="header"
        >
          <div
            class="header-container"
          >
             TodoList 
            <input
              class="header-input"
              data-test="input"
              placeholder="todoItem"
            />
          </div>
        </div>,
        "key": null,
        "patchFlag": 0,
        "props": Object {
          "ref": "VTU_COMPONENT",
        },
        "ref": Object {
          "f": false,
          "i": [Circular],
          "k": undefined,
          "r": "VTU_COMPONENT",
        },
        "scopeId": null,
        "shapeFlag": 4,
        "slotScopeIds": null,
        "ssContent": null,
        "ssFallback": null,
        "staticCount": 0,
        "suspense": null,
        "target": null,
        "targetAnchor": null,
        "transition": null,
        "type": Object {
          "__name": "Header",
          "components": Object {},
          "emits": Array [
            "add",
          ],
          "render": [Function],
          "setup": [Function],
        },
      },
      "suspense": null,
      "suspenseId": 0,
      "type": Object {
        "name": "VTU_ROOT",
        "render": [Function],
      },
      "u": null,
      "uid": 2,
      "um": null,
      "update": [Function],
      "vnode": Object {
        "__v_isVNode": true,
        "__v_skip": true,
        "anchor": null,
        "appContext": Object {
          "app": [Circular],
          "components": Object {
            "transition": Object {
              "name": "transition",
            },
            "transition-group": Object {
              "name": "transition-group",
            },
          },
          "config": Object {
            "compilerOptions": Object {},
            "errorHandler": undefined,
            "globalProperties": Object {},
            "isNativeTag": [Function],
            "optionMergeStrategies": Object {},
            "performance": false,
            "warnHandler": undefined,
          },
          "directives": Object {},
          "emitsCache": WeakMap {},
          "mixins": Array [
            Object {
              "beforeCreate": [Function],
            },
          ],
          "optionsCache": WeakMap {},
          "propsCache": WeakMap {},
          "provides": Object {},
          "reload": [Function],
        },
        "children": null,
        "component": [Circular],
        "dirs": null,
        "dynamicChildren": null,
        "dynamicProps": null,
        "el": <div
          class="header"
        >
          <div
            class="header-container"
          >
             TodoList 
            <input
              class="header-input"
              data-test="input"
              placeholder="todoItem"
            />
          </div>
        </div>,
        "key": null,
        "patchFlag": 0,
        "props": null,
        "ref": null,
        "scopeId": null,
        "shapeFlag": 4,
        "slotScopeIds": null,
        "ssContent": null,
        "ssFallback": null,
        "staticCount": 0,
        "suspense": null,
        "target": null,
        "targetAnchor": null,
        "transition": null,
        "type": Object {
          "name": "VTU_ROOT",
          "render": [Function],
        },
      },
      "withProxy": null,
    },
    "_props": null,
    "_uid": 1,
    "component": [Function],
    "config": Object {
      "compilerOptions": Object {},
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.2.37",
  },
  "__setProps": [Function],
  "componentVM": Object {},
  "isDisabled": [Function],
  "rootVM": Object {},
  "wrapperElement": <div
    class="header"
  >
    <div
      class="header-container"
    >
       TodoList 
      <input
        class="header-input"
        data-test="input"
        placeholder="todoItem"
      />
    </div>
  </div>,
}
`;
